add internal/sys.FileTable

achille-roussel opened the pull request:
As discussed on Slack, this PR brings the file table from https://github.com/achille-roussel/wazero/tree/experimental/sys into upstream Wazero.

I added the data structure as `internal/sys.FileTable`, and I made it a generic type despite the single use in `FSContext` in anticipation that:
- I'm going to bring these changes in https://github.com/achille-roussel/wazero/tree/experimental/sys
- we may have to change the key or value type as we work towards full support of writefs
- we may have multiple instances of it for wasi and gojs (and maybe more with wasi2)

If these assumption turn out not to be true we can revisit and 1) unexport the type and 2) drop the generic type parameters.

Please take a look and let me know if anything needs to be changed!

codefromthecrypt commented on go.mod:
```diff
@@ -1,7 +1,7 @@
 
 // Floor Go version of wazero (current - 2)
-go 1.17
+go 1.18
```
this will hold us back until feb/when go 1.20 is released. are you anticipating keeping this PR open until then, or do you want to work around it for now?

achille-roussel replied:
Ah OK, I may have misunderstood your directions on Slack.

Without this change Go refuses to build the package since generics weren't supported in 1.17. I guess I'll change the type to use concrete implementation and we'll keep copies for now then 👍

codefromthecrypt commented on internal/sys/fs.go:
```diff
@@ -420,11 +392,13 @@ func (c *FSContext) CloseFile(fd uint32) bool {
-	}
+		return true
+	})
+	// TODO: Required to pass TestEmptyFSContext but Reset would be better.
```
a closed context can't be re-opened, why would we have a TODO to use reset?

achille-roussel replied:
I assumed it could, but if it cannot there's no need for this comment indeed 👍

codefromthecrypt commented on internal/sys/fs_test.go:
```diff
@@ -62,12 +62,12 @@ func TestNewFSContext(t *testing.T) {
 			require.NoError(t, err)
```
probably in this file we need to add a test that FDs are re-used. This way a small change to the file table or internals don't invalidate the overall goal which is fd re-use.

ex. 4 == open(f1) 5 == open(f2) 6 == open(f1)  close(4) 4 == open(f2)

codefromthecrypt commented on internal/sys/file_table.go:
```diff
@@ -0,0 +1,141 @@
+
+// Reset clears the content of the table.
+func (t *FileTable) Reset() {
+	for i := range t.masks {
```
do you want to clear the references or truncate the slices instead? e.g. t.masks = t.masks[:]

achille-roussel replied:
I think the convention on `Reset` methods is usually that they don't release the internal state. The data structure is in an "empty" state after clearing the bit masks so I think this is the intended behavior.

codefromthecrypt replied:
this was mainly an idea to get less code. ack it is different to retain the underlying slice vs retain the count of masks (ex clear it as you are). sgtm

codefromthecrypt commented on internal/sys/file_table.go:
```diff
@@ -0,0 +1,141 @@
+// f might return false to interupt the iteration.
+func (t *FileTable) Range(f func(uint32, *FileEntry) bool) {
+	for i, mask := range t.masks {
+		if mask != 0 {
```
nit: prefer early continues vs deep nesting

codefromthecrypt commented on internal/sys/file_table.go:
```diff
@@ -0,0 +1,141 @@
+// Grow ensures that t has enough room for n files, potentially reallocating the
+// internal buffers if their capacity was too small to hold this many files.
+func (t *FileTable) Grow(n int) {
+	if n = (n*8 + 7) / 8; n > len(t.masks) {
```
maybe comment this arithmetic?

codefromthecrypt commented on internal/sys/file_table.go:
```diff
@@ -0,0 +1,141 @@
+// Grow ensures that t has enough room for n files, potentially reallocating the
+// internal buffers if their capacity was too small to hold this many files.
+func (t *FileTable) Grow(n int) {
+	if n = (n*8 + 7) / 8; n > len(t.masks) {
```
since this file looks benchmark optimized you can consider shifting vs multiplying by powers of 2

e.g. `n << 3 // n * 8`

achille-roussel replied:
The Go compiler will do arithmetic operations already (e.g. n*8 is translated to n<<3). The `Grow` method is not intended to be called frequently since the capacity grows exponentially, its cost is amortized over repeated calls to `Insert`.

Unless we have a performance profile showing the optimization would be worth making I'd rather keep the arithmetic in a form that is more readable to human maintainers if that's OK with you.

codefromthecrypt replied:
this was more a nit as I noticed you had some benchmarks. there are other code where we found this improved the results, but agree that "grow" is indeed not one of those frequently called things and thanks for pointing it out!

codefromthecrypt commented on internal/sys/file_table.go:
```diff
@@ -0,0 +1,141 @@
+func (t *FileTable) Insert(file *FileEntry) (fd uint32) {
+	offset := 0
+insert:
+	// TODO: this loop could be made a lot more efficient using vectorized
```
does go support vectorized operations now or in the near future? If so, maybe not a TODO

achille-roussel replied:
I'll change this to just a note rather than a TODO.

codefromthecrypt commented on internal/sys/file_table.go:
```diff
@@ -0,0 +1,141 @@
+
+import "math/bits"
+
+// FileTable is a data structure mapping 32 bit file descriptor to file objects.
```
move the heft of this to /RATIONALE.md then link back to it. e.g. See /RATIONALE.md. this way the godoc can focus on what it is doing overall vs how and why it was made the way it was.

codefromthecrypt reviewed (approved):
Looks nice and professional. I have some feedback which are polish in nature!

achille-roussel commented:
Thanks for the feedback @codefromthecrypt, I believe that I addressed or followed up on all the comments you left.

I also realized that the email address that I had signed the commits with was outdated, so I rewrote the commit history to change signatures to the correct address.

Please let me know if you have any other feedback!

codefromthecrypt commented:
your feedback on feedback was all reasonable. thanks a lot @achille-roussel!
